README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# estatapi - 政府統計の総合窓口（e-Stat）のAPIを使うためのRパッケージ

<!-- badges: start -->

[![R build
status](https://github.com/yutannihilation/estatapi/workflows/R-CMD-check/badge.svg)](https://github.com/yutannihilation/estatapi/actions)
[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/estatapi)](https://CRAN.R-project.org/package=estatapi)
<!-- badges: end -->


*English version of README is
[here](https://github.com/yutannihilation/estatapi/blob/master/README.en.md)*

## e-Stat APIとは

e-Statは日本の統計情報が集まっているポータルサイトです。e-StatにはAPIが用意されていて、[ウェブサイト](http://www.e-stat.go.jp/api/)上でアカウントを登録すると使えるようになります。

**このサービスは、政府統計総合窓口(e-Stat)のAPI機能を使用していますが、サービスの内容は国によって保証されたものではありません。**

## 事前準備

[利用ガイド](http://www.e-stat.go.jp/api/api-guide/)に従ってアプリケーションIDを取得してください。APIにアクセスする際は、`appId`というパラメータに取得したアプリケーションIDを指定します。

## インストール

estatapiはCRAN上に公開されているパッケージなので、通常通りインストールできます。

``` r
install.packages("estatapi")
```

開発版をインストールするには`devtools::install_github()`を使ってください。

``` r
devtools::install_github("yutannihilation/estatapi")
```

## 使い方

現在、このパッケージではバージョン2.0のうち4つをサポートしています。各APIの詳しい解説やパラメータの指定の仕方、返ってくる結果の意味は、公式ドキュメントを参照してください。

  - [統計表情報取得](http://www.e-stat.go.jp/api/e-stat-manual/#api_2_1):
    提供されている統計表を検索します。
  - [メタ情報取得](http://www.e-stat.go.jp/api/e-stat-manual/#api_2_2):
    統計データのメタ情報を取得します。
  - [統計データ取得](http://www.e-stat.go.jp/api/e-stat-manual/#api_2_3):
    統計データを取得します。
  - [データカタログ情報取得](http://www.e-stat.go.jp/api/e-stat-manual/#api_2_6):
    統計表ファイル（Excel、CSV、PDF）および統計データベースの情報を取得します。

### 統計表情報取得（`estat_getStatsList()`）

提供されている統計表を検索します。この関数は、結果を`tbl_df`（dplyrの`data.frame`。`data.frame`とほぼ同じように扱える）として返します。

例えば、「チョコレート」というキーワードを含む統計を検索するときは`searchWord`という引数にキーワードを指定して、以下のようにします。

``` r
appId <- "XXXXXXXXX"
```

``` r
library(estatapi)
#> このサービスは、政府統計総合窓口(e-Stat)のAPI機能を使用していますが、サービスの内容は国によって保証されたものではありません。

estat_getStatsList(appId = appId, searchWord = "チョコレート")
#> # A tibble: 241 x 22
#>    `@id` STAT_NAME GOV_ORG STATISTICS_NAME TITLE CYCLE SURVEY_DATE OPEN_DATE
#>  * <chr> <chr>     <chr>   <chr>           <chr> <chr> <chr>       <chr>    
#>  1 0000… 全国物価統計調査… 総務省  平成9年全国物価統計調査 大… 価格分布… -     199711      2007-09-…
#>  2 0000… 全国物価統計調査… 総務省  平成9年全国物価統計調査 小… 価格分布… -     199711      2007-09-…
#>  3 0000… 全国物価統計調査… 総務省  平成9年全国物価統計調査 消… 指数銘柄… -     199711      2007-09-…
#>  4 0000… 全国物価統計調査… 総務省  平成9年全国物価統計調査 消… 指数銘柄… -     199711      2007-09-…
#>  5 0000… 全国物価統計調査… 総務省  平成9年全国物価統計調査 消… 品目・銘… -     199711      2007-09-…
#>  6 0000… 全国物価統計調査… 総務省  平成9年全国物価統計調査 消… 品目・銘… -     199711      2007-09-…
#>  7 0000… 全国物価統計調査… 総務省  平成9年全国物価統計調査 消… 品目・銘… -     199711      2007-09-…
#>  8 0000… 全国物価統計調査… 総務省  平成9年全国物価統計調査 消… 品目・銘… -     199711      2007-09-…
#>  9 0003… 工業統計調査… 経済産業省… 工業統計調査 確報 平成25… 品目編 … 年次  201301-201… 2015-05-…
#> 10 0003… 工業統計調査… 経済産業省… 工業統計調査 確報 平成26… 品目編 … 年次  201401-201… 2016-04-…
#> # … with 231 more rows, and 14 more variables: SMALL_AREA <chr>,
#> #   COLLECT_AREA <chr>, MAIN_CATEGORY <chr>, SUB_CATEGORY <chr>,
#> #   OVERALL_TOTAL_NUMBER <chr>, UPDATED_DATE <chr>, TABULATION_CATEGORY <chr>,
#> #   TABULATION_SUB_CATEGORY1 <chr>, DESCRIPTION <chr>, TABLE_CATEGORY <chr>,
#> #   TABLE_NAME <chr>, TABULATION_SUB_CATEGORY2 <chr>,
#> #   TABLE_SUB_CATEGORY1 <chr>, TABULATION_SUB_CATEGORY3 <chr>
```

ここで、`STAT_NAME`や`GOV_ORG`は人間が読みやすい形式のラベルになっていますが、
プログラム中で扱う場合はコードのままの方が都合がいいこともあります。そのときは`.use_label
= FALSE`を指定してください。

``` r
estat_getStatsList(appId = appId, searchWord = "チョコレート", .use_label = FALSE)
#> # A tibble: 241 x 22
#>    `@id` STAT_NAME GOV_ORG STATISTICS_NAME TITLE CYCLE SURVEY_DATE OPEN_DATE
#>  * <chr> <chr>     <chr>   <chr>           <chr> <chr> <chr>       <chr>    
#>  1 0000… 00200572  00200   平成9年全国物価統計調査 大… 009   -     199711      2007-09-…
#>  2 0000… 00200572  00200   平成9年全国物価統計調査 小… 009   -     199711      2007-09-…
#>  3 0000… 00200572  00200   平成9年全国物価統計調査 消… 007   -     199711      2007-09-…
#>  4 0000… 00200572  00200   平成9年全国物価統計調査 消… 008   -     199711      2007-09-…
#>  5 0000… 00200572  00200   平成9年全国物価統計調査 消… 009   -     199711      2007-09-…
#>  6 0000… 00200572  00200   平成9年全国物価統計調査 消… 010   -     199711      2007-09-…
#>  7 0000… 00200572  00200   平成9年全国物価統計調査 消… 011   -     199711      2007-09-…
#>  8 0000… 00200572  00200   平成9年全国物価統計調査 消… 012   -     199711      2007-09-…
#>  9 0003… 00550010  00550   工業統計調査 確報 平成25… 1-05  年次  201301-201… 2015-05-…
#> 10 0003… 00550010  00550   工業統計調査 確報 平成26… 1-05  年次  201401-201… 2016-04-…
#> # … with 231 more rows, and 14 more variables: SMALL_AREA <chr>,
#> #   COLLECT_AREA <chr>, MAIN_CATEGORY <chr>, SUB_CATEGORY <chr>,
#> #   OVERALL_TOTAL_NUMBER <chr>, UPDATED_DATE <chr>, TABULATION_CATEGORY <chr>,
#> #   TABULATION_SUB_CATEGORY1 <chr>, DESCRIPTION <chr>, TABLE_CATEGORY <chr>,
#> #   TABLE_NAME <chr>, TABULATION_SUB_CATEGORY2 <chr>,
#> #   TABLE_SUB_CATEGORY1 <chr>, TABULATION_SUB_CATEGORY3 <chr>
```

### メタ情報取得（`estat_getMetaInfo()`）

統計データのメタ情報を取得します。この関数は、結果を`list`として返します。`list`の各要素が、それぞれのデータ項目についてのメタ情報を含んだ`tbl_df`になっています。

例えば、`0003103532`というIDの統計に関するメタ情報を取得するには、`statsDataId`という引数にIDを指定して、以下のようにします。

``` r
meta_info <- estat_getMetaInfo(appId = appId, statsDataId = "0003103532")

names(meta_info)
#> [1] "tab"    "cat01"  "cat02"  "area"   "time"   ".names"

meta_info$cat01
#> # A tibble: 703 x 5
#>    `@code`   `@name`                            `@level` `@unit`  `@parentCode`
#>  * <chr>     <chr>                              <chr>    <chr>    <chr>        
#>  1 000100000 世帯数分布（抽出率調整）           1        一万分比 <NA>         
#>  2 000200000 集計世帯数                         1        世帯     <NA>         
#>  3 000300000 世帯人員                           1        人       <NA>         
#>  4 000400000 18歳未満人員                       2        人       000300000    
#>  5 000500000 65歳以上人員                       2        人       000300000    
#>  6 000600000 65歳以上無職者人員                 3        人       000500000    
#>  7 000700000 有業人員                           1        人       <NA>         
#>  8 000800000 世帯主の年齢                       1        歳       <NA>         
#>  9 000900000 持家率                             1        ％       <NA>         
#> 10 001000000 家賃・地代を支払っている世帯の割合 1        ％       <NA>         
#> # … with 693 more rows
```

### 統計データ取得（`estat_getStatsData()`）

統計データを取得します。この関数は、結果をメタ情報と紐づけて`tbl_df`として返します。

必ず指定しなくてはいけないのは`appId`と`statsDataId`だけですが、それだけだとデータがかなり大きくなって取得に時間がかかります。`cdCat01`（分類事項01）などを指定して必要な項目だけに絞ることをおすすめします。他に絞り込みに指定できるパラメータについては公式ドキュメントを参照してください。

``` r
estat_getStatsData(
  appId = appId,
  statsDataId = "0003103532",
  cdCat01 = c("010800130","010800140")
)
#> Fetching record 1-17792... (total: 17792 records)
#> # A tibble: 17,792 x 13
#>    tab_code 表章項目 cat01_code `品目分類（27年改定）`… cat02_code 世帯区分 area_code
#>  * <chr>    <chr>    <chr>      <chr>            <chr>      <chr>    <chr>    
#>  1 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#>  2 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#>  3 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#>  4 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#>  5 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#>  6 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#>  7 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#>  8 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#>  9 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#> 10 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#> # … with 17,782 more rows, and 6 more variables: 地域区分 <chr>,
#> #   time_code <chr>, `時間軸（月次）` <chr>, unit <chr>, value <dbl>,
#> #   annotation <chr>
```

`limit`で取得する最大のレコード数を、`startPosition`で取得を始めるレコードの位置を指定することもできます。とりあえず少しだけ抜き出して見たい場合や、少しずつデータを取ってきたい場合にはこれらのパラメータが便利です。

``` r
estat_getStatsData(
  appId = appId, statsDataId = "0003103532", cdCat01 = c("010800130","010800140"),
  limit = 3
)
#> Fetching record 1-3... (total: 17792 records)
#> # A tibble: 3 x 13
#>   tab_code 表章項目 cat01_code `品目分類（27年改定）`… cat02_code 世帯区分 area_code
#> * <chr>    <chr>    <chr>      <chr>            <chr>      <chr>    <chr>    
#> 1 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#> 2 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#> 3 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#> # … with 6 more variables: 地域区分 <chr>, time_code <chr>,
#> #   `時間軸（月次）` <chr>, unit <chr>, value <dbl>, annotation <chr>

estat_getStatsData(
  appId = appId, statsDataId = "0003103532", cdCat01 = c("010800130","010800140"),
  startPosition = 101,
  limit = 3
)
#> Fetching record 101-103... (total: 17792 records)
#> # A tibble: 3 x 13
#>   tab_code 表章項目 cat01_code `品目分類（27年改定）`… cat02_code 世帯区分 area_code
#> * <chr>    <chr>    <chr>      <chr>            <chr>      <chr>    <chr>    
#> 1 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#> 2 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#> 3 01       金額     010800130  352 チョコレート 03         二人以上の世帯… 00000    
#> # … with 6 more variables: 地域区分 <chr>, time_code <chr>,
#> #   `時間軸（月次）` <chr>, unit <chr>, value <dbl>, annotation <chr>
```

### データカタログ情報取得（`estat_getDataCatalog()`）

統計表ファイル（Excel、CSV、PDF）および統計データベースの情報を取得できます。

このAPIはファイルのURLを返すだけなので、そのままRで処理することは難しいかもしれません。

``` r
catalog1 <- estat_getDataCatalog(appId = appId, searchWord = "チョコレート", dataType = c("PDF", "XLS"))

catalog1[1, c("@id", "STAT_NAME", "TABLE_NAME", "SURVEY_DATE", "TABLE_SUB_CATEGORY1", "DATASET_NAME", "NAME", "LANDING_PAGE", "URL", "FORMAT")] %>%
  glimpse
#> Rows: 1
#> Columns: 10
#> $ `@id`               <chr> "000000701890"
#> $ STAT_NAME           <chr> "全国物価統計調査"
#> $ TABLE_NAME          <chr> "業態別価格分布－全国，都市階級，都道府県"
#> $ SURVEY_DATE         <chr> "1997"
#> $ TABLE_SUB_CATEGORY1 <chr> "0104チョコレート"
#> $ DATASET_NAME        <chr> "平成9年全国物価統計調査_大規模店舗編_1997年"
#> $ NAME                <chr> "価格分布_1_業態別価格分布－全国，都市階級，都道府県_0104チョコレート"
#> $ LANDING_PAGE        <chr> "https://www.e-stat.go.jp/stat-search/files?layo…
#> $ URL                 <chr> "https://www.e-stat.go.jp/stat-search/file-downl…
#> $ FORMAT              <chr> "XLS"
```


NEWS.md
# estatapi (development version)

# estatapi 0.4.0

## Major Changes

* Use e-Stat API version 3.0.

# estatapi 0.3.0

## Major Changes

* `estat_getStatsData()` and `estat_getStatsList()` has been overhauled. Accordingly, each function
  returns the result with different colnames than before. I'm sorry, but this change was inevitable
  to fix bugs like [#7](https://github.com/yutannihilation/estatapi/issues/7)... For the details,
  please see [this gist](https://gist.github.com/yutannihilation/e89e3cab6212aa9a390c2c831310ddae).
* Treat most of parameters as formal arguments. Please be careful if you use unnamed arguments.
* Use e-Stat API version 2.1, which is released on Jul 14, 2016.
* `estat_getStatsData()` automatically repeats fetching records when the records are more than the
  number that can be fetched at one time (=100,000 records). You can stop this behaviour by specifying
  `limit` or `.fetch_all` argument.
* Add `getDataCatalog()`, which finds statistical dataset files (Excel, CSV and PDF) and databases.

## Minor Changes

* Show license notice on startup.


tests/testthat.R
library(testthat)
library(estatapi)

test_check("estatapi")


tests/testthat/test-api.R
context("API with mocks")

dummy_res <- structure(
  list(
    url = "",
    status_code = 200L,
    headers = structure(
      list(
        `content-type` = "application/json;charset=utf-8",
        `content-encoding` = "gzip"
      ),
      .Names = c(
        "content-type",
        "content-encoding"
      ),
      class = c("insensitive", "list")
    ),
    content = raw(0)
  ),
  .Names = c(
    "url",
    "status_code",
    "headers",
    "content"
  ),
  class = "response"
)

test_that("estat_getStatsList processes the API response as expected", {
  with_mock(
    `httr::GET` = function(...) {
      purrr::update_list(dummy_res, content = readRDS("content_getStatsList.rds"))
    },
    expect_equivalent(
      estat_getStatsList(
        appId = "XXXX",
        searchWord = "CD",
        limit = 3
      ),
      readRDS("result_getStatsList.rds")
    )
  )
})

test_that("estat_getStatsList (w/o label) processes the API response as expected", {
  with_mock(
    `httr::GET` = function(...) {
      purrr::update_list(dummy_res, content = readRDS("content_getStatsList.rds"))
    },
    expect_equivalent(
      estat_getStatsList(
        appId = "XXXX",
        searchWord = "CD",
        limit = 3,
        .use_label = FALSE
      ),
      readRDS("result_getStatsList_wo_label.rds")
    )
  )
})

test_that("estat_getMetaInfo processes the API response as expected", {
  with_mock(
    `httr::GET` = function(...) {
      purrr::update_list(dummy_res, content = readRDS("content_getMetaInfo.rds"))
    },
    expect_equivalent(
      estat_getMetaInfo(appId = "XXXX", statsDataId = "0003065345"),
      readRDS("result_getMetaInfo.rds")
    )
  )
})

test_that("estat_getStatsData processes the API response as expected", {
  with_mock(
    `estatapi:::estat_getStatsDataCount` = function(...) 40014,
    `estatapi::estat_getMetaInfo` = function(...) readRDS("result_getMetaInfo.rds"),
    `httr::GET` = function(...) {
      res <- purrr::update_list(dummy_res,
        content = readRDS("content_getStatsData.rds")
      )
      res$headers$`content-type` <- "text/plain; charset=utf-8"
      res
    },
    {
      expect_equivalent(
        estat_getStatsData(
          appId = "XXXX",
          statsDataId = "0003065345",
          cdCat01 = c("008", "009", "010"),
          limit = 3
        ),
        readRDS("result_getStatsData.rds")
      )
    }
  )
})


test_that("estat_getDataCatalog processes the API response as expected", {
  with_mock(
    `httr::GET` = function(...) {
      purrr::update_list(dummy_res, content = readRDS("content_getDataCatalog.rds"))
    },
    expect_equivalent(
      estat_getDataCatalog(
        appId = "XXXX",
        searchWord = "\u30c1\u30e7\u30b3\u30ec\u30fc\u30c8",
        dataType = c("PDF", "XLS")
      ),
      readRDS("result_getDataCatalog.rds")
    )
  )
})


test_that("estat_getDataCatalog (w/o label) processes the API response as expected", {
  with_mock(
    `httr::GET` = function(...) {
      purrr::update_list(dummy_res, content = readRDS("content_getDataCatalog.rds"))
    },
    expect_equivalent(
      estat_getDataCatalog(
        appId = "XXXX",
        searchWord = "\u30c1\u30e7\u30b3\u30ec\u30fc\u30c8",
        dataType = c("PDF", "XLS"),
        .use_label = FALSE
      ),
      readRDS("result_getDataCatalog_wo_label.rds")
    )
  )
})


tests/testthat/test-api-real.R
context("API with real request")


# Constants ---------------------------------------------

EXPECTED_NAMES <- c(
  "tab_code", "\u8868\u7ae0\u9805\u76ee",
  "cat01_code", "\u54c1\u76ee\u5206\u985e\uff0827\u5e74\u6539\u5b9a\uff09",
  "cat02_code", "\u4e16\u5e2f\u533a\u5206",
  "area_code", "\u5730\u57df\u533a\u5206",
  "time_code", "\u6642\u9593\u8ef8\uff08\u6708\u6b21\uff09",
  "unit",
  "value",
  "annotation"
)

STATS_DATA_ID <- "0003103532"

CD_CAT01_SMALL <- c("010800130", "010800140")

CD_CAT01_LARGE <- c(
  "000100000", "000200000", "000300000", "000400000", "000500000",
  "000600000", "000700000", "000800000", "000900000", "001000000",
  "001100000", "010000000", "010100000", "010110001", "010120000",
  "010120010", "010120020"
)


# Util Functions ----------------------------------------

wrap_api_func <- function(fun, ...) {
  skip_on_cran()
  skip_on_ci()

  if (!"estatapi" %in% keyring::key_list()$service) {
    skip('appId is not available; Set the key by keyring::key_set("estatapi")')
  }

  appId <- keyring::key_get("estatapi")
  purrr::partial(fun, appId = appId)
}

check_df <- function(d, expected_names = EXPECTED_NAMES, expected_count = NULL) {
  # Confirm there are no duplicated rows
  expect_equal(nrow(d), nrow(dplyr::distinct(d)))

  # Confirm the column names are the same as expected
  expect_equal(names(d), expected_names)

  if (!is.null(expected_count)) {
    # Confirm the number of rows is the same as expected
    expect_equal(nrow(d), expected_count)
  } else {
    expect_true(nrow(d) > 0)
  }
}


# Tests with <1000000 records -----------------------------------------------

test_that("estat_getStatsData with <100000 records works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_SMALL
  )

  check_df(d)
})

test_that("estat_getStatsData with <100000 records and limit works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_SMALL,
    limit = 10
  )

  check_df(d, expected_count = 10)
})

test_that("estat_getStatsData with <100000 records and limit works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_SMALL,
    limit = 10
  )

  check_df(d, expected_count = 10)
})


test_that("estat_getStatsData with <100000 records and startPosition works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_SMALL,
    startPosition = 10001
  )

  total_number <- wrap_api_func(estat_getStatsDataCount)(statsDataId = STATS_DATA_ID, cdCat01 = CD_CAT01_SMALL)
  check_df(d, expected_count = total_number - 10001 + 1)
})

test_that("estat_getStatsData with <100000 records and startPosition and limit works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_SMALL,
    startPosition = 10001,
    limit = 10
  )

  check_df(d, expected_count = 10)
})


# Tests with >1000000 records -----------------------------------------------

test_that("estat_getStatsData with >100000 records works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_LARGE
  )

  check_df(d)
})

test_that("estat_getStatsData with >100000 records and limit >100000 works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_LARGE,
    limit = 110000
  )

  check_df(d, expected_count = 110000)
})

test_that("estat_getStatsData with >100000 records and limit <100000 works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_LARGE,
    limit = 90000
  )

  check_df(d, expected_count = 90000)
})

test_that("estat_getStatsData with >100000 records and startPosition works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_LARGE,
    startPosition = 10001
  )

  total_number <- wrap_api_func(estat_getStatsDataCount)(statsDataId = STATS_DATA_ID, cdCat01 = CD_CAT01_LARGE)
  check_df(d, expected_count = total_number - 10001 + 1)
})

test_that("estat_getStatsData with >100000 records and startPosition and limit >1000000 works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_LARGE,
    startPosition = 10001,
    limit = 100800
  )

  check_df(d, expected_count = 100800)
})

test_that("estat_getStatsData with >100000 records and startPosition and limit <1000000 works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_LARGE,
    startPosition = 10001,
    limit = 90000
  )

  check_df(d, expected_count = 90000)
})


# Test English -----------------------------------------------

test_that("estat_getStatsData with <100000 records in English works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = "0003036792",
    cdTab = "2",
    cdArea = "13A01",
    cdCat01 = "7009",
    lang = "E"
  )

  check_df(d,
    expected_names = c(
      "tab_code", "Tabulated variable",
      "cat01_code", "Items",
      "area_code", "AREA",
      "time_code", "Time",
      "unit",
      "value",
      "annotation"
    )
  )
})


R/util.R
#' Util

ESTAT_API_URL <- "http://api.e-stat.go.jp/"

#' @title e-Stat API
#'
#' @description Get Statistical Something From e-Stat API
#'
#' @param path
#'   API endpoint.
#' @param appId
#'   Application ID.
#' @param ...
#'   Other parameters.
#'
#' @export
estat_api <- function(path, appId, ...) {
  # convert params like cdCat01=c("001","002") to cdCat01="001,002"
  query <- flatten_query(list(appId = appId, ...))

  res <- httr::GET(
    ESTAT_API_URL,
    config = httr::add_headers(`Accept-Encoding` = "gzip"),
    path = path,
    query = query
  )

  httr::warn_for_status(res)

  media <- httr::parse_media(res$headers$`content-type`)

  # A result of gettStatsData, getMetaInfo, getDataCatalog and getStatsList is JSON.
  # That of getSimpleStatsData is CSV (but the media type is "text/plain").
  if (identical(media$type, "application") && identical(media$subtype, "json")) {
    parse_result_json(res)
  } else if (identical(media$type, "text") && identical(media$subtype, "plain")) {
    parse_result_csv(res)
  } else {
    stop("unknown media type: ", res$headers$`content-type`, call. = FALSE)
  }
}

flatten_query <- function(x) {
  x <- purrr::compact(x)
  x <- purrr::map(x, ~ paste0(as.character(.), collapse = ","))

  # Ignore duplicated elements
  x[unique(names(x))]
}

as_flattened_character <- function(x, .use_label = TRUE) {
  x <- purrr::map(x, try_dollar, .use_label = .use_label)
  x <- purrr::flatten(x)
  x <- purrr::map(x, as.character)

  x
}

# 1) scalar value  ->  return x as it is
# 2) list with "$" element
#   2-1) if .use_label is TRUE  ->  return the value of "$"
#   2-2) if .use_label is FALSE ->  return the other value
# 3) list without "$" element  ->  return x as it is (x will be flattened outside of this function)
try_dollar <- function(x, .use_label = TRUE) {
  if (!is.list(x)) {
    return(x)
  }
  if (!"$" %in% names(x)) {
    return(x)
  }

  if (.use_label) {
    x[["$"]]
  } else {
    x[[which(names(x) != "$")]]
  }
}

get_class_info <- function(class_obj) {
  meta_ids <- purrr::map_chr(class_obj, "@id")
  meta_names <- purrr::map_chr(class_obj, "@name")

  class_info <- purrr::map(class_obj, ~ dplyr::bind_rows(.$CLASS))
  names(class_info) <- meta_ids

  purrr::update_list(class_info, .names = tibble::tibble(id = meta_ids, name = meta_names))
}

parse_result_json <- function(res) {
  result_parsed <- httr::content(res)

  if (result_parsed[[1]]$RESULT$STATUS != 0) {
    stop(result_parsed[[1]]$RESULT$ERROR_MSG)
  }

  result_parsed
}

parse_result_csv <- function(res) {
  result_text <- httr::content(res)

  # value column has `-` to represent missing values
  suppressWarnings(
    result_parsed <- readr::read_csv(result_text,
      col_types = readr::cols(
        value = readr::col_number(),
        .default = readr::col_character()
      )
    )
  )

  # read_csv() returns a spec_tbl_df, which cannot be handled by dplyr::bind_rows().
  # So, convert it to a tibble explicitly.
  tibble::as_tibble(result_parsed)
}


R/getMetaInfo.R
#' getMetaInfo API
#'
#' Get meta-information about the statistical dataset via e-Stat API.
#'
#' @param appId
#'   Application ID.
#' @param statsDataId
#'   ID of the statistical dataset.
#' @param lang
#'   Language of data. `"J"`(Japanese) or `"E"`(English).
#' @param ...
#'   Other parameters.
#'
#' @seealso
#' <http://www.e-stat.go.jp/api/e-stat-manual3-0/#api_2_2>
#' <http://www.e-stat.go.jp/api/e-stat-manual3-0/#api_3_3>
#'
#' @examples
#' \dontrun{
#' estat_getMetaInfo(appId = "XXXX", statsDataId = "0003065345")
#' }
#'
#' @export
estat_getMetaInfo <- function(appId, statsDataId, lang = c("J", "E"), ...) {
  lang <- match.arg(lang)
  j <- estat_api("rest/3.0/app/json/getMetaInfo", appId = appId, statsDataId = statsDataId, lang = lang, ...)

  get_class_info(j$GET_META_INFO$METADATA_INF$CLASS_INF$CLASS_OBJ)
}


R/getStatsData.R
#' getStatsData API
#'
#' Get some statistical data via e-Stat API.
#'
#' @param appId
#'   Application ID.
#' @param statsDataId
#'   ID of the statistical dataset.
#' @param startPosition
#'   Index of the first record to get.
#' @param limit
#'   Max number of records to get.
#' @param lang
#'   Language of the data. `"J"`(Japanese) or `"E"`(English).
#' @param .fetch_all
#'   Whether to fetch all records when the number of records is larger than
#'   100,000.
#' @param ...
#'   Other parameters.
#' \itemize{
#'  \item `lvTab`:
#'    Level of the meta-information. The format can be `X` or `X-Y`, `-X` and `X-`.
#'  \item `cdTab`:
#'    Code(s) of the meta-infomation items to select. The format can be a character vector (`c("001", "002")`) or
#'    a chraracter of values and commas (`"001,002"`).
#'  \item `cdTabFrom`:
#'    The code of the first meta-information item to select.
#'  \item `cdTabTo`:
#'    The code of the last meta-information item to select.
#'  \item `lvTime`:
#'     Level of the time to select. The format is the same as `lvTab`
#'  \item `cdTime`
#'     Time(s) to select. The format is the same way like `cdTab`
#'  \item `cdTimeFrom`:
#'    The first time to select. The format is the same way like `cdTabFrom`
#'  \item `cdTimeTo`:
#'    The last time to select. The format is the same way like `cdTabTo`
#'  \item `lvArea`:
#'     Level of the area to select. The format is the same as `lvTab`
#'  \item `cdArea`
#'     Code(s) of the Area to select. The format is the same way like `cdTab`
#'  \item `cdAreaFrom`:
#'    The code of the first area to select. The format is the same way like `cdTabFrom`
#'  \item `cdAreaTo`:
#'    The code of the last area to select. The format is the same way like `cdTabTo`
#'  \item `lvCat01`, `cdCat01`, `cdCat01From`, `cdCat01To`, ...:
#'    The same way like above.
#' }
#'
#' @seealso
#' <http://www.e-stat.go.jp/api/e-stat-manual3-0/#api_2_3>
#' <http://www.e-stat.go.jp/api/e-stat-manual3-0/#api_3_4>
#'
#' @examples
#' \dontrun{
#' # fetch all data, which may take time
#' estat_getStatsData(
#'   appId = "XXXX",
#'   statsDataId = "0003065345"
#' )
#'
#' # fetch data up to 10 records
#' estat_getStatsData(
#'   appId = "XXXX",
#'   statsDataId = "0003065345",
#'   limit = 10
#' )
#'
#' # fetch data up to 100,000 records (max number of records available at once)
#' estat_getStatsData(
#'   appId = "XXXX",
#'   statsDataId = "0003065345",
#'   .fetch_all = FALSE
#' )
#'
#' # fetch all data in the specifed category
#' estat_getStatsData(
#'   appId = "XXXX",
#'   statsDataId = "0003065345",
#'   cdCat01 = c("008", "009", "010")
#' )
#' }
#'
#' @export
estat_getStatsData <- function(appId, statsDataId,
                               startPosition = NULL,
                               limit = NULL,
                               lang = c("J", "E"),
                               .fetch_all = TRUE,
                               ...) {
  lang <- match.arg(lang)

  total_number <- estat_getStatsDataCount(appId, statsDataId, lang = lang, ...)
  ranges <- calc_ranges(startPosition, limit, total_number, .fetch_all)

  result <- list()

  for (i in seq_len(nrow(ranges))) {
    cur_limit <- ranges$limits[i]
    cur_start <- ranges$starts[i]

    message(sprintf(
      "Fetching record %.0f-%.0f... (total: %.0f records)\n",
      cur_start, cur_start + cur_limit - 1, total_number
    ))

    j <- estat_api("rest/3.0/app/getSimpleStatsData",
      appId = appId,
      statsDataId = statsDataId,
      startPosition = format(cur_start, scientific = FALSE),
      limit = format(cur_limit, scientific = FALSE),
      lang = lang,
      metaGetFlg = "N",
      sectionHeaderFlg = "2", # No header
      ...
    )

    result[[i]] <- j
  }

  dplyr::bind_rows(result)
}

estat_getStatsDataCount <- function(appId, statsDataId, ...) {
  j <- estat_api("rest/3.0/app/json/getStatsData",
    appId = appId,
    statsDataId = statsDataId,
    metaGetFlg = "N",
    cntGetFlg = "Y",
    ...
  )

  as.numeric(j$GET_STATS_DATA$STATISTICAL_DATA$RESULT_INF$TOTAL_NUMBER)
}

calc_ranges <- function(startPosition,
                        limit,
                        record_count,
                        .fetch_all,
                        .max_records_at_once = 100000) {
  ranges <- list()

  if (is.null(startPosition)) {
    startPosition <- 1
  }

  if (is.null(limit)) {
    endPosition <- record_count
  } else {
    endPosition <- min(startPosition + limit - 1, record_count)
  }

  if (.fetch_all) {
    ranges$starts <- seq(from = startPosition, to = endPosition, by = .max_records_at_once)
    ranges$limits <- rep(.max_records_at_once, length(ranges$starts))
    # treat a fraction
    fraction <- (endPosition - startPosition + 1) %% .max_records_at_once
    if (fraction != 0) {
      ranges$limits[length(ranges$limits)] <- fraction
    }
  } else {
    ranges$starts <- startPosition
    ranges$limits <- min(startPosition + .max_records_at_once - 1, endPosition)
  }

  tibble::tibble(!!!ranges)
}
